### PR TITLE
[v7r2] Use a conda environment when building the docs

### PR DESCRIPTION
**Pull request by chrisburr**
I think it would be nicer to use conda to build the docs so no modules have to be mocked and minimise the number of places the dependencies are listed. I don't think I have a way of testing this so there might be some more changes needed.

I've also renamed `.readthedocs.yml` as according to [here](https://docs.readthedocs.io/en/stable/config-file/index.html) the `.yml` name has been deprecated in favour of `.yaml`

BEGINRELEASENOTES

*docs
CHANGE: Use a conda environment for building the docs

ENDRELEASENOTES

### CODE REVIEW

**Comment by andresailer** [comment]
You can make your own dirac readthedocs copy and push it there, or let me push it to mine
https://diracsailer.readthedocs.io/

**Comment by andresailer** [comment]
There is now also the option to build PRs on readthedocs
https://docs.readthedocs.io/en/latest/pull-requests.html

**Comment by chrisburr** [comment]
> There is now also the option to build PRs on readthedocs
> https://docs.readthedocs.io/en/latest/pull-requests.html

That seems nice, shall we try replacing `tests/checkDocs.sh` with that or should I fix the GitHub actions build as well?

**Comment by andresailer** [comment]
Let's first see how that RTD PR check looks like, and if that fails for warnings as well

The checkDocs also has this nice check
https://github.com/DIRACGrid/DIRAC/blob/daa908158832081b32ecf64fcbe313c7ffae8065/tests/checkDocs.sh#L11

**Comment by andresailer** [comment]
close/open to see if this triggers on RTD

File: docs/diracdoctools/__init__.py [hunk 1]
```diff
@@ -8,17 +8,4 @@
...
-                       'gfal2',
-                       'git',
-                       'irods',
-                       'lcg_util',
-                       'MySQLdb',
-                       'pylab',
-                       'pythonjsonlogger',
-                       'stomp',
-                       ]
+DIRAC_DOC_MOCK_LIST = []
```
Comment (andresailer) [inline docs/diracdoctools/__init__.py:11]: Maybe keep the list available for people inheriting this list and not building on conda

Comment (chrisburr) [reply docs/diracdoctools/__init__.py:11]: Do we want to support that? This mocking has always felt like a hack to workaround the fact that it was hard to install the dependencies.

Comment (andresailer) [reply docs/diracdoctools/__init__.py:11]: Do we want to stop supporting this, is the question.
Also this https://github.com/DIRACGrid/DIRAC/blob/integration/docs/source/DeveloperGuide/CodeDocumenting/index.rst#building-the-documentation needs an update

**Comment by andresailer** [comment]
testing: https://readthedocs.org/projects/diracsailer/builds/13190631/

**Comment by chrisburr** [comment]
Thanks @andresailer! I already have a build here: https://readthedocs.org/projects/cburrdirac/builds/13190420/

Do you happen to know how to deal with these issues?

```log
checkouts/latest/docs/source/AdministratorGuide/CommandReference/dirac-install.rst:9: WARNING: Unexpected indentation.
checkouts/latest/src/DIRAC/Core/Utilities/Graphs/GraphUtilities.py:docstring of matplotlib.ticker.Formatter.fix_minus:1: WARNING: Unknown interpreted text role "rc".
checkouts/latest/src/DIRAC/Core/Utilities/Graphs/GraphUtilities.py:docstring of matplotlib.ticker.Formatter.fix_minus:1: WARNING: Unknown interpreted text role "rc".
checkouts/latest/src/DIRAC/Core/Utilities/Graphs/GraphUtilities.py:docstring of matplotlib.ticker.ScalarFormatter.set_powerlimits:9: WARNING: Unknown interpreted text role "rc".
checkouts/latest/src/DIRAC/Core/Utilities/Graphs/GraphUtilities.py:docstring of matplotlib.ticker.ScalarFormatter.set_useLocale:3: WARNING: Unknown interpreted text role "rc".
checkouts/latest/src/DIRAC/Core/Utilities/Graphs/GraphUtilities.py:docstring of matplotlib.ticker.ScalarFormatter.set_useMathText:5: WARNING: Unknown interpreted text role "rc".
checkouts/latest/src/DIRAC/Core/Utilities/Graphs/GraphUtilities.py:docstring of matplotlib.ticker.ScalarFormatter.set_useOffset:10: WARNING: Unknown interpreted text role "rc".
checkouts/latest/src/DIRAC/DataManagementSystem/private/HttpStorageAccessHandler.py:docstring of http.server.BaseHTTPRequestHandler.send_error:5: WARNING: Unexpected indentation.
checkouts/latest/src/DIRAC/DataManagementSystem/private/HttpStorageAccessHandler.py:docstring of http.server.BaseHTTPRequestHandler.send_error:6: WARNING: Block quote ends without a blank line; unexpected unindent.
checkouts/latest/src/DIRAC/DataManagementSystem/private/HttpStorageAccessHandler.py:docstring of http.server.BaseHTTPRequestHandler.send_error:7: WARNING: Inline emphasis start-string without end-string.
```

**Comment by chrisburr** [comment]
> Let's first see how that RTD PR check looks like, and if that fails for warnings as well

See here: https://github.com/chrisburr/DIRAC/pull/8

It's not quite as nice for displaying the warnings as you have to scroll to the area between `reading sources...` and `writing output...` but I think it's good enough. Thoughts?

**Comment by andresailer** [comment]
you could try with updating this https://github.com/DIRACGrid/DIRAC/blob/daa908158832081b32ecf64fcbe313c7ffae8065/docs/source/conf.py#L301-L302 to python3 and the same matplotlib version as is used, or "stable"?

Not sure what dirac-install is in this build, so not sure which line needs fixing...

**Comment by andresailer** [comment]
When everything fails: add the dirac modules to https://github.com/DIRACGrid/DIRAC/blob/daa908158832081b32ecf64fcbe313c7ffae8065/docs/docs.conf#L20 which will prevent the documentation for the problematic functions to be included

**Comment by chrisburr** [comment]
Thanks for  the help, `no_inherited_members` ended up being needed.

I've now updated the documentation and made it so that the docs are build in parallel where possible.

File: tests/checkDocs.sh [hunk 2]
```diff
@@ -1,21 +0,0 @@
...
-
-sudo apt-get install graphviz;
-cd docs;
-# need to have this folder in the PYTHONPATH so we can find diracdoctools
-export PYTHONPATH=$PWD:$PYTHONPATH
-
-SPHINXOPTS=-wsphinxWarnings make htmlall || { echo "Failed to build documentation, check for errors above" ; exit 1; }
-
-# check for :param / :return in html, points to faulty syntax, missing empty lines, etc.
-grep --color -nH -e :param -e :return -r build/html/CodeDocumentation/ >> sphinxWarnings
```
[the remark this answers is not in the document]
Comment (chrisburr) [reply tests/checkDocs.sh:11]: Can you give me an example of how I can trigger it?

[the remark this answers is not in the document]
Comment (chrisburr) [reply tests/checkDocs.sh:11]: I've figured out how to, now let's see if I can make Sphinx notice...

[the remark this answers is not in the document]
Comment (chrisburr) [reply tests/checkDocs.sh:11]: I'm not proud of it but how about 32f195445:

```log
The HTML pages are in build/html.
build/html/CodeDocumentation/ConfigurationSystem/Client/CSAPI.html:71:<p>:param str siteName FQN of the site (e.g. LCG.CERN.ch)
build/html/CodeDocumentation/ConfigurationSystem/Client/CSAPI.html:72::param str ceName: FQN of the CE (e.g. ce503.cern.ch)
build/html/CodeDocumentation/ConfigurationSystem/Client/CSAPI.html:73::param dict optionsDict: optional dictionary of options
build/html/CodeDocumentation/ConfigurationSystem/Client/CSAPI.html:74::returns: S_OK/S_ERROR structure</p>
Return code from grep --color -nH -e :param -e :return -r build/html/CodeDocumentation/ was 0
This means :param or :return in the html and points to faulty syntax, missing empty lines, etc.
make: *** [Makefile:54: html] Error 1
```

File: docs/diracdoctools/cmd/commandReference.py [hunk 3]
```diff
@@ -116,16 +119,15 @@ def createFiles(self, sectionDict):
...
-      scriptName = os.path.basename(script)
-      if scriptName.endswith('.py'):
-        scriptName = scriptName[:-3]
-      scriptName = scriptName.replace("_", "-")
-      prefix = sectionDict[PREFIX].lower()
-      prefix = prefix + '_' if prefix else ''
-      if self.createScriptDocFiles(script, sectionPath, scriptName, referencePrefix=prefix) and \
-         scriptName not in commandList:
+    for future in futures:
+      scriptName, createdScriptDocs = future.result()
```
Comment (andresailer) [inline docs/diracdoctools/cmd/commandReference.py:129]: This is a left over after the tidying, isn't it? `future.result()` isn't a tuple, but just the `True/False` from `createScriptDocFiles`, right?

Comment (chrisburr) [reply docs/diracdoctools/cmd/commandReference.py:129]: Nice spot! It still needs to be a tuple for the `scriptName not in commandList`.

It should be fixed now, it's being built in https://readthedocs.org/projects/cburrdirac/builds/13199424/

File: tests/checkDocs.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (andresailer) [reply tests/checkDocs.sh:11]: Yes, very nice! As long as it fails :)

File: docs/diracdoctools/Utilities.py [hunk 4]
```diff
@@ -129,3 +130,22 @@ def makeLogger(name):
...
   return logging.getLogger(name)
+
+
+def registerValidatingExitHandler():
+  """Registers an exit handler which checks for errors after the build completes"""
+  def check():
+    """"""
+    cmd = [
+        "grep", "--color", "-nH", "-e", ":param",
+        "-e", ":return", "-r", "build/html/CodeDocumentation/"
```
Comment (andresailer) [inline docs/diracdoctools/Utilities.py:141]: from your build
> The HTML pages are in _build/html.
> Updating searchtools for Read the Docs search... grep: build/html/CodeDocumentation/: No such file or directory


```suggestion
        "-e", ":return", "-r", "_build/html/CodeDocumentation/"
```

And change the builddir to `_build` in the Makefile

**Comment by atsareg** [comment]
have you converged on this PR ?

File: docs/diracdoctools/Utilities.py (hunk 4 above)
Comment (chrisburr) [reply docs/diracdoctools/Utilities.py]: Fixed now: https://readthedocs.org/projects/cburrdirac/builds/13202691/

I went with crude argument parsing like the other bits in `conf.py` as it currently uses `build/` when running manually.

**Comment by andresailer** [comment]
Someone (@atsareg @fstagni ) needs to activate the read the docs PR integration in the diracgrid github side, I think?

**Comment by atsareg** [comment]
So, this is no more a WIP PR and I can merge it to activate. Is there anything else to activate it ?

**Comment by chrisburr** [comment]
When testing I only had to enable it on the readthedocs side: https://docs.readthedocs.io/en/latest/pull-requests.html#preview-documentation-from-pull-requests

**Comment by andresailer** [comment]
https://docs.readthedocs.io/en/latest/pull-requests.html
I activated the thing on the rtd side, but I don't see PRs popping up in the builds of rtd yet.

**Comment by andresailer** [comment]
@atsareg 
https://readthedocs.org/dashboard/dirac/integrations/62426/
Could you try clicking on "re-sync" webhook?

**Comment by atsareg** [comment]
Result:


    Your project, DIRAC, needs to be reconfigured in order to continue building automatically after January 31st, 2019. For more information, see our documentation on webhook integrations.
    Webhook activation failed. Make sure you have the necessary permissions.

**Comment by chrisburr** [comment]
Maybe the easiest way of dealing with this it is to temporarily make either me or Andre admins in the DIRACGrid organisation?

**Comment by andresailer** [comment]
that or screen sharing. (and now I deleted the so far working webhook T_T)

**Comment by chrisburr** [comment]
Alternatively @atsareg, can you check at https://readthedocs.org/accounts/social/connections/ if your GitHub account is linked to readthedocs? If it is, maybe try disconnecting/re-connecting it before clicking on the resync button again.

**Comment by andresailer** [comment]
image
looking better

**Comment by chrisburr** [comment]
> Command killed due to excessive memory consumption

I read somewhere that conda builds are ran in VMs with more resources available. I'm retrying but as the docs haven't had any issues building in my fork perhaps we should merge to see if that triggers readthedocs to switch to the bigger VMs.

If that fails I'll send them an email asking they enable the `mamba` feature flag: https://docs.readthedocs.io/en/stable/guides/conda.html#troubleshooting

**Comment by andresailer** [comment]
no 😢 
> Command killed due to excessive memory consumption

Maybe the multi-threading?

**Comment by chrisburr** [comment]
I don't see anything relevant and the same build passed on my fork every time it's been retried: https://readthedocs.org/projects/cburrdirac/builds/13197843/

**Comment by andresailer** [comment]
weird also that the last one timed out at 1021 seconds, but the final result was success
The only difference in the log I was was
```
  matplotlib.font_manager:     WARN: Matplotlib is building the font cache; this may take a moment.
```
in the PR one

**Comment by andresailer** [comment]
one more try

**Comment by andresailer** [comment]
The build time is more than the 15 minutes (900s) usually allowed. Now it is around 1000, before it was around 400
https://docs.readthedocs.io/en/stable/builds.html

**Comment by chrisburr** [comment]
> I'm almost certain that the issue is that my fork is being router to the big nodes while yours/DIRAC is still trying to use the small ones, see: https://blog.readthedocs.com/better-conda-support/

While drafting an email I stumbled upon proof that my suspicion was correct: https://github.com/readthedocs/readthedocs.org/blob/7450d6d36376f3372173729604a2edad239deb85/readthedocs/builds/tasks.py#L85-L104

I'll still send the request for CONDA_USES_MAMBA but I think enabling conda on the default branch will fix the issue.

**Comment by chrisburr** [comment]
And I've also opened an issue upstream to fix or document this quirk: https://github.com/readthedocs/readthedocs.org/issues/8004

**Comment by andresailer** [comment]
Yes, this does imply there is a different node is used for the builds. It is probably worth merging and seeing what happens, but I am not fully sure it will actually pick the big nodes if we just switch the default, maybe still request that in your email?

https://github.com/readthedocs/readthedocs.org/blob/7450d6d36376f3372173729604a2edad239deb85/readthedocs/builds/tasks.py#L106-L107

It depends what the filter actually means, but if that is `successful` builds that are checked for conda, then enabling on the default branch is not going to be enough, unless we accidentally build something successfully? Maybe after "wiping" some versions?

**Comment by chrisburr** [comment]
I've pushed a branch to the DIRACGrid repo that successfully builds with conda so it should be okay: https://readthedocs.org/projects/dirac/builds/13216715/

To get "external" builds (PRs?) to work we might also have to use conda on the "latest" branch. As we're about to release v7r2 I don't think this is much of a problem.

**Comment by chrisburr** [comment]
As usual, `mamba` to the rescue: https://readthedocs.org/projects/dirac/builds/13220033/

**Comment by atsareg** [comment]
So, I can merge it now ?

**Review by andresailer — APPROVED** [review]

**Review by atsareg — APPROVED** [review]

**Pull request merged**